### PR TITLE
Close WAL block file before deleting everything

### PR DESCRIPTION
**Pull request by kostya9**
**What this PR does**: Close WAL block file before deleting everything

Also added some other files closing that I found during a look around the code.

PS: I not proficient in Go, so I could have misunderstood, so please check if it looks good

**Which issue(s) this PR fixes**:
Attempt to fix https://github.com/grafana/tempo/issues/2034
Fixes https://github.com/grafana/tempo/issues/2153
**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: tempodb/encoding/vparquet/wal_block.go [hunk 1]
```diff
@@ -446,6 +446,7 @@ func (b *walBlock) Iterator() (common.Iterator, error) {
 }
 
 func (b *walBlock) Clear() error {
+	b.file.Close()
```
Comment (kostya9) [inline tempodb/encoding/vparquet/wal_block.go:449]: This is the thing that I think may fix the original issue. `b.writer` is not closed too, but as far as I see, that it is based on the `file`, so there is no reason to double close

File: tempodb/encoding/vparquet/wal_block.go [hunk 2]
```diff
@@ -446,6 +446,9 @@ func (b *walBlock) Iterator() (common.Iterator, error) {
 }
 
 func (b *walBlock) Clear() error {
+	if b.file != nil {
+		b.file.Close()
```
Comment (kostya9) [inline tempodb/encoding/vparquet/wal_block.go:450]: This is the thing that I think may fix the original issue. b.writer is not closed too, but as far as I see, that it is based on the file, so there is no reason to double close

Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go:450]: Here, I debugged test `TestWalDropsZeroLength`. Not sure whether the test is simulating real-world workload, but the file is still open when we call `CutCompleteTraces`

image

File: tempodb/backend/local/local.go [hunk 3]
```diff
@@ -131,6 +131,7 @@ func (rw *Backend) Read(ctx context.Context, name string, keypath backend.KeyPat
 
 	stat, err := f.Stat()
 	if err != nil {
+		f.Close()
```
Comment (kostya9) [inline tempodb/backend/local/local.go:134]: I am not sure about this one - should we close if open succeeded and stat failed?

Comment (joe-elliott) [reply tempodb/backend/local/local.go:134]: how about just deferring close? i think that's ok for read only files.

File: tempodb/encoding/vparquet/wal_block.go (hunk 2 above)
Comment (joe-elliott) [reply tempodb/encoding/vparquet/wal_block.go:450]: i definitely agree with this change, but let's check and return the error from `b.file.Close()`.

should we run `os.RemoveAll()` even if `b.file.Close()` fails? it might be the safer thing to do

File: tempodb/encoding/vparquet/wal_block.go [hunk 4]
```diff
@@ -460,6 +463,8 @@ func (b *walBlock) FindTraceByID(ctx context.Context, id common.ID, _ common.Sea
 			}
 
 			r := parquet.NewReader(file)
+			defer r.Close()
```
Comment (joe-elliott) [inline tempodb/encoding/vparquet/wal_block.go:466]: should we also defer close the file above? and all of the files returned from `page.file()` calls?

**Review by joe-elliott — COMMENTED** [review]
thanks for taking a look at this :+1:

File: tempodb/backend/local/local.go (hunk 3 above)
Comment (kostya9) [reply tempodb/backend/local/local.go:134]: Cant defer close in the higher scope because f is returned from the function. I assume the caller expects an opened file

File: tempodb/encoding/vparquet/wal_block.go (hunk 4 above)
Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go:466]: `parquet.File` unfortunately does not implement a Close function, so we can not close it. I noticed that when I was figuring out what is going on and logged an issue - https://github.com/grafana/tempo/issues/2153

File: tempodb/encoding/vparquet/wal_block.go (hunk 2 above)
Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go]: Added error aggregation. I agree, I think it's safer to remove everything rather than not, so did that.

**Review by joe-elliott — COMMENTED** [review]
apologies. i had a pending review that i did not submit from a couple days back.

File: tempodb/encoding/vparquet/wal_block.go (hunk 4 above)
Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go:473]: Sounds interesting.

What do you think of wrapping the parquet.File with another type implements Close()? I feel like maybe that would be a more intuitive design. Also, it could later be used in other call sites theoretically (which would probably also need to close the file at some point).

Comment (joe-elliott) [reply tempodb/encoding/vparquet/wal_block.go:473]: so like:
```
type pageFile struct {
  parquet.File
  os.File
}
```
then the caller can interact with the parquet.file and eventually close the os.file? i'm good on that

Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go:508]: Added closing the file opened in `file()`. It resulted in a little bit more changes that I anticipated, but we should close everything opened in the wal_block module now

Comment (kostya9) [reply tempodb/encoding/vparquet/wal_block.go:508]: Ah, did all that work, but didnt call the finel `iterator.Close()`. Fixed that

File: CHANGELOG.md [hunk 5]
```diff
@@ -19,6 +19,8 @@
 * [BUGFIX] Correctly coalesce trace level data when combining Parquet traces. [#2095](https://github.com/grafana/tempo/pull/2095) (@joe-elliott)
 * [BUGFIX] Unescape query parameters in AWS Lambda to allow TraceQL queries to work. [#2114](https://github.com/grafana/tempo/issues/2114) (@joe-elliott)
 * [CHANGE] Pad leading zeroes in span id to always be 16 chars [#2062](https://github.com/grafana/tempo/pull/2062) (@ie-pham)
+* [BUGFIX] Fix float/int comparisons in TraceQL. [#2139](https://github.com/grafana/tempo/issues/2139) (@joe-elliott)
```
Comment (joe-elliott) [inline CHANGELOG.md:22]: we redid the changelog for 2.0.1. can you merge main or rebase and clean this up?

**Review by joe-elliott — APPROVED** [review]
@mapno I'm good on this, but before merging can you take a quick look? this does a lot of subtle changes with regard to the wal and searching. i'm concerned we may see some strange issues crop up b/c we relied on these files being opened, but i think explicitly closing these files is the right direction.

don't feel the need to do a line by line review (unless you want). mainly flagging you due to the changes along the SearchTags and TagValues paths.

@kostya9 thanks for the persistence and level of detail on this! after @mapno takes a look we will merge.

File: CHANGELOG.md (hunk 5 above)
Comment (kostya9) [reply CHANGELOG.md]: Sure, rebased and fixed this

**Review by mapno — APPROVED** [review]
LGTM. This change touches a lot of code, so it's difficult to follow all the `close()` calls. Tests should cover most edge cases. Thanks!

**Comment by kostya9** [comment]
I see failures in the generator storage tests, from my analysis, the generator doesn't use the code I changed here. Are these tests flaky, or is there a chance something is broken?

**Comment by joe-elliott** [comment]
> I see failures in the generator storage tests, from my analysis, the generator doesn't use the code I changed here. Are these tests flaky, or is there a chance something is broken?

Yeah. It's a flakey test. Will rerun.

**Comment by joe-elliott** [comment]
Thanks for the fixes @kostya9!
